lib/Target/SubstraitPB/Import.cpp

```diff
@@ -0,0 +1,101 @@
+//===-- Import.cpp - Import protobuf to Substrait dialect -------*- C++ -*-===//
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "structured/Target/SubstraitPB/Import.h"
+
+#include "mlir/IR/BuiltinOps.h"
+#include "mlir/IR/ImplicitLocOpBuilder.h"
+#include "mlir/IR/OwningOpRef.h"
+#include "structured/Dialect/Substrait/IR/Substrait.h"
+
+#include <google/protobuf/text_format.h>
+#include <substrait/plan.pb.h>
+
+using namespace mlir;
+using namespace mlir::substrait;
+
+namespace pb = google::protobuf;
+
+namespace {
+
+// Forward declaration for the import function of the given message type.
+//
+// We need one such function for most message types that we want to import. The
+// forward declarations are necessary such all import functions are available
+// for the definitions indepedently of the order of these definitions. The
+// message type passed to the function (specified by `MESSAGE_TYPE`) may be
+// different than the one it is responsible for: often the target op type
+// (specified by `OP_TYPE`) depends on a nested field value (such as `oneof`)
+// but the import logic needs the whole context; the message that is passed in
+// is the most deeply nested message that provides the whole context.
+#define DECLARE_IMPORT_FUNC(MESSAGE_TYPE, ARG_TYPE, OP_TYPE)                   \
+  static FailureOr<OP_TYPE> import##MESSAGE_TYPE(                              \
+      ImplicitLocOpBuilder builder, const ::substrait::ARG_TYPE &message);
+
+DECLARE_IMPORT_FUNC(Plan, Plan, PlanOp)
+DECLARE_IMPORT_FUNC(PlanRel, PlanRel, PlanRelOp)
+
+static FailureOr<PlanOp> importPlan(ImplicitLocOpBuilder builder,
+                                    const ::substrait::Plan &message) {
+  const ::substrait::Version &version = message.version();
+  auto planOp = builder.create<PlanOp>(
+      version.major_number(), version.minor_number(), version.patch_number(),
+      version.git_hash(), version.producer());
+  planOp.getBody().push_back(new Block());
+
+  for (int i = 0; i < message.relations_size(); i++) {
```

Comment:
Nit: https://llvm.org/docs/CodingStandards.html#don-t-evaluate-end-every-time-through-a-loop , could you use a range based for here?

Reply:
Yep, range-based works!